src/librbd/io/ReadResult.cc

```diff
@@ -191,26 +189,49 @@ void ReadResult::C_ObjectReadRequest::finish(int r) {
     r = 0;
   }
   if (r >= 0) {
-    ldout(cct, 10) << " got " << extent_map
-                   << " for " << buffer_extents
-                   << " bl " << bl.length() << dendl;
+    uint64_t object_len = 0;
     aio_completion->lock.lock();
-    if (!extent_map.empty()) {
+    for (auto& extent: extents) {
+      ldout(cct, 10) << " got " << extent.extent_map
+                     << " for " << extent.buffer_extents
+                     << " bl " << extent.bl.length() << dendl;
+
       aio_completion->read_result.m_destriper.add_partial_sparse_result(
-        cct, bl, extent_map, object_off, buffer_extents);
-    } else {
-      // handle the case where a sparse-read wasn't issued
-      aio_completion->read_result.m_destriper.add_partial_result(
-        cct, std::move(bl), buffer_extents);
+              cct, extent.bl, extent.extent_map, extent.offset,
+              extent.buffer_extents);
+
+      object_len += extent.length;
     }
     aio_completion->lock.unlock();
-
     r = object_len;
   }
 
   aio_completion->complete_request(r);
 }
 
+ReadResult::C_ObjectReadMergedExtents::C_ObjectReadMergedExtents(
+        CephContext* cct, ReadExtents* extents, Context* on_finish)
+        : cct(cct), extents(extents), on_finish(on_finish) {
+}
+
+void ReadResult::C_ObjectReadMergedExtents::finish(int r) {
+  if (r == 0) {
```

Comment:
I haven't fully traced the callers to this yet, but I think read IO usually returns the number of bytes read as a `r` value, so shouldn't this be `r >= 0`?